examples/example_poly.rs

```diff
@@ -0,0 +1,164 @@
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE
+// or https://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or https://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+//
+// Copyright (c) ZK-GARAGE. All rights reserved.
+
+//! PLONK Example
+
+use ark_bls12_381::{Bls12_381, Fr as BlsScalar};
+use ark_ec::TEModelParameters;
+use ark_ed_on_bls12_381::{
+    EdwardsParameters as JubJubParameters, Fr as JubJubScalar,
+};
+use ark_ff::PrimeField;
+use ark_poly::polynomial::univariate::DensePolynomial;
+use ark_poly_commit::{sonic_pc::SonicKZG10, PolynomialCommitment};
+use plonk::error::to_pc_error;
+use plonk_core::circuit::{verify_proof, Circuit};
+use plonk_core::constraint_system::StandardComposer;
+use plonk_core::error::Error;
+use plonk_core::prelude::*;
+use rand_core::OsRng;
+
+fn main() -> Result<(), Error> {
+    // Implements a circuit that defines the following function:
+    // -------------------
+    // def f(x, y, z):
+    //   if x == 1:
+    //     return y*z
+    //   return 2y - z
+    // -------------------
+    // where x,y,z are inputs and r is the output
+
+    #[derive(derivative::Derivative)]
+    #[derivative(Debug(bound = ""), Default(bound = ""))]
+    pub struct TestCircuit<F, P>
+    where
+        F: PrimeField,
+        P: TEModelParameters<BaseField = F>,
+    {
+        x: F,
+        y: F,
+        z: F,
+        r: F,
+        dummy: P::ScalarField,
+    }
+
+    impl<F, P> Circuit<F, P> for TestCircuit<F, P>
+    where
+        F: PrimeField,
+        P: TEModelParameters<BaseField = F>,
+    {
+        const CIRCUIT_ID: [u8; 32] = [0xff; 32];
+
+        fn gadget(
+            &mut self,
+            composer: &mut StandardComposer<F, P>,
+        ) -> Result<(), Error> {
+            let x = composer.add_input(self.x);
+            let y = composer.add_input(self.y);
+            let z = composer.add_input(self.z);
+            let r = composer.add_input(self.r);
+
+            let one = composer.add_input(F::one());
+            let two = composer.add_input(F::from(2u64));
+            let minus_1 = composer.add_input(-F::one());
+
+            // Define 2y
+            let two_y = composer.arithmetic_gate(|gate| {
+                gate.witness(y, two, None).mul(F::one())
+            });
+
+            // Define -z
+            let minus_z = composer.arithmetic_gate(|gate| {
+                gate.witness(z, minus_1, None).mul(F::one())
+            });
+
+            // Define 2y-z
+            let two_y_minus_z = composer.arithmetic_gate(|gate| {
+                gate.witness(two_y, minus_z, None).add(F::one(), F::one())
```

Comment:
Since `2` is a constant and the subsctraction is just an addition with an implied `-1` constant we can eliminate the inputs: `one`, `two` and `minus_1` as well as the previous gates that compute `two_y` and `minus_z`. We can fit all the computation in just one gate:
```
let two_y_minus_z = composer.arithmetic_gate(|gate| {
    gate.witness(y, z, None).add(F::from(2u64), -F::one())
});
```